bigtable-hbase-1.x-parent/bigtable-hbase-1.x/src/test/java/com/google/cloud/bigtable/hbase/TestRetryBehavior.java

```diff
@@ -0,0 +1,297 @@
+    public String toString() {
+      return name;
+    }
+  }
+
+  @Parameterized.Parameters(name = "{0}")
+  public static Object[] getOps() {
+    return new Object[] {
+      new TestOp("single-get", (table) -> table.get(newGet("a"))),
+      new TestOp("single-put", (table) -> table.put(newPut("some-key"))),
+      new TestOp("scan", (table) -> Lists.newArrayList(table.getScanner(new Scan()))),
+      new TestOp("multi-get", (table) -> table.get(Arrays.asList(newGet("a"), newGet("b")))),
+      new TestOp("multi-put", (table) -> table.put(Arrays.asList(newPut("a"), newPut("b"))))
+    };
+  }
+
+  public TestRetryBehavior(TestOp testOp) {
+    this.testOp = testOp;
+  }
+
+  @Before
+  public void setup() throws IOException {
+    fakeServer =
+        TestServerBuilder.newInstance()
+            .intercept(
+                new ServerInterceptor() {
+                  @Override
+                  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+                      ServerCall<ReqT, RespT> serverCall,
+                      Metadata metadata,
+                      ServerCallHandler<ReqT, RespT> serverCallHandler) {
+                    callCount.incrementAndGet();
+
+                    // record the deadline
+                    @Nullable
+                    Duration duration =
+                        Optional.ofNullable(Context.current().getDeadline())
+                            .map(d -> Duration.ofMillis(d.timeRemaining(TimeUnit.MILLISECONDS)))
+                            .orElse(null);
+                    remainingDurations.add(duration);
+
+                    // execute test specific callback
+                    rpcRunnable.get().accept(serverCall);
+
+                    // Don't really care about client requests, just blackhole them
+                    // this will prevent any invocations of the stub
+                    @SuppressWarnings("unchecked")
+                    ServerCall.Listener<ReqT> blackhole = Mockito.mock(ServerCall.Listener.class);
+                    return blackhole;
+                  }
+                })
+            .addService(new BigtableGrpc.BigtableImplBase() {})
+            .buildAndStart();
+  }
+
+  @After
+  public void teardown() {
+    fakeServer.shutdownNow();
+  }
+
+  private Configuration createBaseConfig() {
+    Configuration config = BigtableConfiguration.configure("fake-project", "fake-instance");
+    config.set(BigtableOptionsFactory.PROJECT_ID_KEY, "fake-project");
+    config.set(BigtableOptionsFactory.INSTANCE_ID_KEY, "fake-project");
+    config.set(BigtableOptionsFactory.BIGTABLE_DATA_CHANNEL_COUNT_KEY, "1");
+    config.set(
+        BigtableOptionsFactory.BIGTABLE_EMULATOR_HOST_KEY, "localhost:" + fakeServer.getPort());
+    config.set(BigtableOptionsFactory.ADDITIONAL_RETRY_CODES, "ABORTED");
+
+    config.setLong(BigtableOptionsFactory.INITIAL_ELAPSED_BACKOFF_MILLIS_KEY, DELAY.toMillis());
+
+    setOperationTimeout(config, OPERATION_TIMEOUT);
```

Comment:
Should we also add tests to verify default timeouts?

Reply:
The default timeouts are too long and would cause the tests be very slow